【Hackathon 9th Sprint No.9】feat: implement ES(t) macro/micro cross-validation and refactor analysis utilities

Dayuxiaoshui opened the pull request:
This commit implements the Error-aware Speedup Score (ES_t) metric from Section 3.2.2 of the technical report (arXiv:2510.24035), along with the mathematical proofs from Appendix B and C that establish the sample-level validity of both S_t and ES_t metrics.

Key Features:
=============

1. Appendix B Implementation - Sample-level proof for S_t:
   - Micro-level calculation: geometric mean of rectified speedups for all samples
   - Macro-level calculation: S_t = α^λ · β^(ληp) · b^(1-λ)
   - Cross-validation: both methods produce identical results, proving S_t is equivalent to the geometric mean of sample-level rectified speedups

2. Appendix C Implementation - Sample-level proof for ES_t:
   - Micro-level calculation: geometric mean of error-aware rectified speedups
   - Macro-level calculation: ES_t = α^λ · β^(ληp) · γ_t^(1-λ)
   - Dynamic penalty factor: γ_t = b^(sum(π_c * indicator(t < c)))
   - Cross-validation: validates that ES_t is the geometric mean of error-aware rectified speedups, where failure samples use type-specific dynamic penalties instead of fixed penalty b

3. Error-aware design (Section 3.2.2):
   - Error type classification: c=1 (accuracy), c=2 (runtime crash), c=3 (compile failure)
   - Tiered tolerance rules: t≥1 tolerates accuracy errors, t≥2 tolerates runtime crashes, t≥3 tolerates all errors
   - Dynamic penalty γ_t adapts based on error type distribution and tolerance level

4. Independent verification script:
   - verify_macro_params.py: calculates and prints all macro parameters (alpha, beta, gamma, lambda, eta, pi) independently
   - Enables validation of plot_ESt results by computing each parameter separately

5. Mandatory validation mechanism:
   - plot_ESt.py: enforces macro/micro result matching before adoption
   - Rejects results if validation fails, ensuring calculation correctness

6. Code refactoring for maintainability:
   - macro_statistics.py: dedicated module for macro parameter calculations
   - Each parameter has independent function (alpha, beta, gamma, lambda, eta, pi)
   - Reduced nesting levels in analysis_util.py by extracting helper functions
   - Simplified scan_all_folders and added .txt file support
   - Improved code organization following software engineering best practices

Technical Details:
==================

- Micro calculation: processes each sample individually, applies rectified speedup rules, then computes geometric mean
- Macro calculation: uses aggregated statistics (correct count, speedup distributions, error type proportions) to compute expected values
- Validation: compares micro and macro results with tolerance threshold (1e-6)
- All calculations verified against real benchmark data (118 samples)

Files Changed:
==============
- graph_net/analysis_util.py: refactored with helper functions, integrated macro_statistics module, reduced nesting, simplified scan_all_folders
- graph_net/macro_statistics.py: new module for macro parameter calculations
- graph_net/plot_ESt.py: added mandatory macro/micro validation
- graph_net/verify_macro_params.py: new independent verification script

All code passes pre-commit checks, compiles successfully, and has been validated with real benchmark data.

### PR Category
<!-- One of [ New Sample | Feature Enhancement | Bug Fix | Other ] -->


### Description
<!-- Describe what you’ve done -->

Dayuxiaoshui committed:
feat: implement ES(t) macro/micro cross-validation and refactor analysis utilities

This commit implements the Error-aware Speedup Score (ES_t) metric from
Section 3.2.2 of the technical report (arXiv:2510.24035), along with the
mathematical proofs from Appendix B and C that establish the sample-level
validity of both S_t and ES_t metrics.

Key Features:
=============

1. Appendix B Implementation - Sample-level proof for S_t:
   - Micro-level calculation: geometric mean of rectified speedups for all samples
   - Macro-level calculation: S_t = α^λ · β^(ληp) · b^(1-λ)
   - Cross-validation: both methods produce identical results, proving S_t
     is equivalent to the geometric mean of sample-level rectified speedups

2. Appendix C Implementation - Sample-level proof for ES_t:
   - Micro-level calculation: geometric mean of error-aware rectified speedups
   - Macro-level calculation: ES_t = α^λ · β^(ληp) · γ_t^(1-λ)
   - Dynamic penalty factor: γ_t = b^(sum(π_c * indicator(t < c)))
   - Cross-validation: validates that ES_t is the geometric mean of
     error-aware rectified speedups, where failure samples use type-specific
     dynamic penalties instead of fixed penalty b

3. Error-aware design (Section 3.2.2):
   - Error type classification: c=1 (accuracy), c=2 (runtime crash), c=3 (compile failure)
   - Tiered tolerance rules: t≥1 tolerates accuracy errors, t≥2 tolerates
     runtime crashes, t≥3 tolerates all errors
   - Dynamic penalty γ_t adapts based on error type distribution and tolerance level

4. Independent verification script:
   - verify_macro_params.py: calculates and prints all macro parameters
     (alpha, beta, gamma, lambda, eta, pi) independently
   - Enables validation of plot_ESt results by computing each parameter separately

5. Mandatory validation mechanism:
   - plot_ESt.py: enforces macro/micro result matching before adoption
   - Rejects results if validation fails, ensuring calculation correctness

6. Code refactoring for maintainability:
   - macro_statistics.py: dedicated module for macro parameter calculations
   - Each parameter has independent function (alpha, beta, gamma, lambda, eta, pi)
   - Reduced nesting levels in analysis_util.py by extracting helper functions
   - Simplified scan_all_folders and added .txt file support
   - Improved code organization following software engineering best practices

Technical Details:
==================

- Micro calculation: processes each sample individually, applies rectified
  speedup rules, then computes geometric mean
- Macro calculation: uses aggregated statistics (correct count, speedup
  distributions, error type proportions) to compute expected values
- Validation: compares micro and macro results with tolerance threshold (1e-6)
- All calculations verified against real benchmark data (118 samples)

Files Changed:
==============
- graph_net/analysis_util.py: refactored with helper functions, integrated
  macro_statistics module, reduced nesting, simplified scan_all_folders
- graph_net/macro_statistics.py: new module for macro parameter calculations
- graph_net/plot_ESt.py: added mandatory macro/micro validation
- graph_net/verify_macro_params.py: new independent verification script

All code passes pre-commit checks, compiles successfully, and has been
validated with real benchmark data.

lixinqi commented on graph_net/macro_statistics.py:
samples_statistics.py

macro 这一词不好理解，顺带把相关变量名都改一下。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    return gmean(correct_speedups) if correct_speedups else 1.0
+
+
+def calculate_beta(slowdown_speedups: List[float]) -> float:
```
传correct_speedups

函数内用临时变量
`slowdown_speedups =  list(filter(lambda x:x<1, correct_speedups))`

小心维护全系统层面的概念，少一个算一个，能局部处理就不要让所有模块感知。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    Returns:
+        Alpha value (geometric mean), or 1.0 if list is empty
+    """
+    return gmean(correct_speedups) if correct_speedups else 1.0
```
`gmean(correct_speedups) if len(correct_speedups) > 0 else 1.0`

加强语义。各家语言的 bool 强转是一笔烂账，不要轻易触碰。每次都把代码写清楚，坚持最小惊讶原则。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+"""
+
+from scipy.stats import gmean
+from typing import List, Tuple
```
有些 linter 已经提倡直接用 list 和 tuple 了

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    Returns:
+        Beta value (geometric mean), or 1.0 if list is empty
+    """
+    return gmean(slowdown_speedups) if slowdown_speedups else 1.0
```
len(slowdown_speedups) > 0

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    return gmean(slowdown_speedups) if slowdown_speedups else 1.0
+
+
+def calculate_lambda(correct_count: int, total_samples: int) -> float:
```
坚持用 correct_speedups
correct_count由correct_speedups推算得到

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    Returns:
+        Lambda value (fraction of correct samples), or 0.0 if total_samples is 0
+    """
+    return correct_count / total_samples if total_samples > 0 else 0.0
```
else 的值是不是应该为 1.0?
感觉应该宽容点。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    return correct_count / total_samples if total_samples > 0 else 0.0
+
+
+def calculate_eta(correct_negative_speedup_count: int, correct_count: int) -> float:
```
坚持只传 correct_speedups。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+
+
+def calculate_pi(
+    acc_failure_count: int, total_samples: int, correct_count: int
```
acc_failure_count 变量名语义不明确再想想。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+
+    # indicator[0] = 1 if t < 1 else 0 (accuracy errors not tolerated)
+    # indicator[1] = 1 if t < 3 else 0 (runtime/compile errors not tolerated)
+    indicator = [1 if t_key < 1 else 0, 1 if t_key < 3 else 0]
```
过分和论文的错误码耦合在一起了。
实际上用户会自定义错误码，到时这些代码都用不起来。

lixinqi commented on graph_net/macro_statistics.py:
```diff
@@ -0,0 +1,257 @@
+    return pi_0, pi_1
+
+
+def calculate_gamma(t_key: int, pi: Tuple[float, float], fpdb: float = 0.1) -> float:
```
按照计算公式，参数应该是：
```
def calculate_gamma(tolerance: int, get_pi: Callable[int, float], errno_tolerances: list[int], b: float = 0.1) -> float:

```
变量名能不缩写就不缩写，它本身承载了语义。代码绝大部分场合是给人看的，变量名比注释更频繁地被人注意到。

lixinqi commented on graph_net/plot_ESt.py:
这个文件里尽量不要动原来的函数。

新增代码是简单的，删改代码却需要非常谨慎。

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -138,16 +138,66 @@ def main():
         )
-        all_es_scores[folder_name] = es_scores
+
+        # Verify macro/micro consistency
```
新写的一个函数，不应该超过 30 行。
把业务逻辑细节用函数名来合理表达，用函数调用函数的自然嵌套来提供可维护性。

lixinqi commented on graph_net/verify_macro_params.py:
```diff
@@ -0,0 +1,244 @@
+        slowdown_speedups = []
+
+        # Collect statistics for current tolerance using helper function
+        for idx, sample in enumerate(samples):
```
换成 map/reduce/flat_map/filter/group 或者 list comprehension 等形式。

控制自己写结构化逻辑的惯性。结构化编程随意性很大，bug 容易滋生，而上面两种形式约束更强，语义也更强，易于维护。

Dayuxiaoshui committed:
refactor: rename macro to aggregated and improve code quality

This commit refactors the evaluation metrics calculation code with the following improvements:

1. Terminology refactoring: macro -> aggregated
   - Rename macro_statistics.py to samples_statistics.py
   - Rename verify_macro_params.py to verify_aggregated_params.py
   - Update all variable and function names accordingly

2. Code structure improvements
   - Extract verification logic in plot_ESt.py into separate functions
     * compare_single_tolerance_level (12 lines)
     * print_verification_result (1 line)
     * verify_aggregated_micro_consistency (28 lines, meets ≤30 line requirement)
   - Refactor verify_aggregated_params.py to use functional programming style
     * Replace structured loops with list comprehensions
     * Use Counter for error type counting
     * Reduce multiple traversals to single pass where possible

3. Reduce function parameter coupling
   - calculate_beta: derive slowdown_speedups internally from correct_speedups
   - calculate_lambda: derive correct_count internally from correct_speedups
   - calculate_eta: derive statistics internally from correct_speedups

4. Decouple error type handling
   - calculate_pi: accept error_type_counts (dict) instead of hardcoded types
   - calculate_gamma: accept generic parameters (tolerance, get_pi, errno_tolerances)
   - Support user-defined error codes instead of hardcoded error types

5. Code quality improvements
   - Use explicit len() checks instead of implicit boolean conversion
   - Use modern Python type hints (list/tuple instead of typing.List/Tuple)
   - Improve code readability and maintainability

All changes have been verified and pass pre-commit checks.

Dayuxiaoshui committed:
style: apply black formatting to samples_statistics.py and verify_aggregated_params.py

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,300 @@
+
+
+def calculate_pi(
+    error_type_counts: dict[str, int], total_samples: int, correct_speedups: list[float]
```
```python
errno2count: dict[int, int]
```
errno 实际上是需要排序的。str 看不出来。

lixinqi commented on graph_net/analysis_util.py:
```diff
@@ -456,39 +568,36 @@ def calculate_s_scores(
         t_key,
         correct_count,
-        acc_failure_count,
+        error_type_counts,
```
都统一成 `errno2count` 
如果某些地方需要用到 error_type 信息，则给它传一个 error_str4errno 函数

Dayuxiaoshui committed:
refactor: unify error type to errno mapping for better sorting

- Replace error_type_counts (dict[str, int]) with errno2count (dict[int, int])
- Add get_errno_from_error_type() to map error type strings to errno (1, 2, 3)
- Add get_error_type_from_errno() for reverse mapping when error type strings are needed
- Update calculate_pi() to use errno2count and return dict[int, float]
- Update calculate_all_aggregated_parameters() to use errno2count and errno_tolerance_thresholds
- Update analysis_util.py and verify_aggregated_params.py to use errno2count
- Improve code maintainability by using integer errno for sorting and comparison

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    """
+    correct_count = len(correct_speedups)
+    error_count = total_samples - correct_count
+    if error_count == 0:
```
```python
assert error_count == sum(errno2count.values())
```

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+
+def calculate_gamma(
+    tolerance: int,
+    get_pi: Callable[[int], float],
```
这个 fp 的思路非常好。把函数最为第一等公民用，能让calculate_gamma自身的逻辑更为清爽，复用度也会大大提高。

lixinqi replied:
函数名字的语义不足，应该加强语义：pi_value4errno

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+def calculate_gamma(
+    tolerance: int,
+    get_pi: Callable[[int], float],
+    errno_tolerances: list[int],
```
errno_as_tolerances

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+
+    # Calculate indicator for each error type: 1 if not tolerated, 0 if tolerated
+    pi_sum = 0.0
+    for error_type_index in range(len(errno_tolerances)):
```
这里语义不清晰。首先，这里不应该再出现 error_type 概念，全程都应该只有 errno 概念。其次error_type_index 更含糊，这种局部概念会让代码可维护性很差，而且不能简单直接地看出来问题。

lixinqi replied:
应该直接改成
```python
for errno in errno_tolerances:
```

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    # Calculate indicator for each error type: 1 if not tolerated, 0 if tolerated
+    pi_sum = 0.0
+    for error_type_index in range(len(errno_tolerances)):
+        pi_c = get_pi(error_type_index)
```
改成
```python
pi_value = pi_value4errno(errno)
```

lixinqi replied:
重新调整pi_value4errno 输入参数的语义

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    pi_sum = 0.0
+    for error_type_index in range(len(errno_tolerances)):
+        pi_c = get_pi(error_type_index)
+        threshold_c = errno_tolerances[error_type_index]
```
去掉

threshold_c就是 errno

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+        threshold_c = errno_tolerances[error_type_index]
+        # Error type is not tolerated (penalized) when t < threshold
+        indicator = 1 if tolerance < threshold_c else 0
+        pi_sum += pi_c * indicator
```
应该尝试用 list comprehension 来简化
```python
pi_sum = sum(
    pi_value
    for errno in errno_tolerances
    for pi_value in [pi_value4errno(errno)]
    if tolerance < errno
)
```

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    """
+    if len(errno_tolerances) == 0:
+        return b
+
```
```python
if tolerance <= 0:
    return b
```

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    lambda_: float,
+    eta: float,
+    negative_speedup_penalty: float,
+    fpdb: float,
```
都统一写成 b

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+        lambda_: Fraction of correct samples
+        eta: Fraction of slowdown cases within correct samples
+        negative_speedup_penalty: Penalty power p for negative speedup
+        fpdb: Base penalty b for severe errors
```
b： Base penalty for severe errors or accuracy violation

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    return b**pi_sum
+
+
+def calculate_s_t_from_aggregated(
```
这个函数内部应该调用 es 的版本，因为 b 是 gamma 的特例

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    total_samples: int,
+    correct_speedups: list[float],
+    errno2count: dict[int, int],
+    t_key: int,
```
tolerance

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    )
+
+
+def calculate_all_aggregated_parameters(
```
这个名字有点迷惑。尤其是 all 这个词

lixinqi replied:
可以改名成 calculate_es_components_values

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    correct_speedups: list[float],
+    errno2count: dict[int, int],
+    t_key: int,
+    negative_speedup_penalty: float = 0.0,
```
可以简写为 p，解释到为就行。

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    errno2count: dict[int, int],
+    t_key: int,
+    negative_speedup_penalty: float = 0.0,
+    fpdb: float = 0.1,
```
简写为 b

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    negative_speedup_penalty: float = 0.0,
+    fpdb: float = 0.1,
+    pi: dict[int, float] | None = None,
+    errno_tolerance_thresholds: dict[int, int] | None = None,
```
不要出现 thresholds 这样的词，因为它的语义太抽象。你统一用 errno 来表达。

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+    # Convert dictionary-based pi and thresholds to indexed format for calculate_gamma
+    # Create ordered list of errnos for consistent indexing (sorted by errno)
+    errnos = sorted(errno2count.keys())
+    errno_tolerances = [errno_tolerance_thresholds.get(errno, 3) for errno in errnos]
```
从 304行到319行的代码所关注的细节非常底层，与本函数其他部分代码层级格格不入，不利于维护。

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,350 @@
+        "gamma": gamma,
+        "pi": pi,
+        "s_t": s_t,
+        "es_t": es_t,
```
去掉这些字段，让外边的代码来调用calculate_s_t_from_aggregated 和calculate_es_t_from_aggregated
s_t 和 es_t 与其他字段画风差异太大了。其他是“构成”，而 s_t 和 es_t是“整体”了。

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,275 @@
+        )
+
+        # Store state at t=1 using list comprehension
+        if t_key == 1:
```
t_key 都统一为 tolerance

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,275 @@
+            fail_type_at_t1 = [fail_type for _, _, _, fail_type in t1_data]
+
+        # Calculate pi at t=1 using the dedicated function
+        if t_key == 1:
```
这些特判都是不应该出现在这个函数里的。

本函数篇幅太长的原因是因为考虑了过多的细节。正常做法是应该放到子函数里去考虑的。

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,275 @@
+            final_errno2count = errno2count.copy()  # Save for t >= 1
+
+        # Calculate aggregated parameters
+        if total_samples > 0:
```
这些琐碎的细节不应该出现在这里。

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,275 @@
+)
+
+
+def calculate_aggregated_parameters(
```
这个名字的语义太弱了，看不出里边在做啥

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,275 @@
+    print(f"Aggregated Parameter Verification Complete")
+    print(f"{'='*80}\n")
+
+    return results
```
既然本函数最重要的逻辑是出一个报表，为什么不直接写成一句话：
```python
return {
    tolerance: {
        "alpha": get_alpha(tolerance),
        "beta": get_beta(tolerance),
        "gamma": get_gamma(tolerance),
        "lambda": get_lambda(tolerance),
        "eta": get_eta(tolerance),
        "pi": get_pi(tolerance),
        "expected_s": get_expected_s(tolerance),
        "expected_es": get_expected_es(tolerance),
        "correct_count": get_correct_count(tolerance),
        "num_total_samples": num_total_samples,
        "num_correct_speedups": getnum_correct_speedups(tolerance),
        "num_slowdown": get_num_slowdown(tolerance),
    }
    for tolerance in range(-10, num_error_types + 2)
}
```
然后把上面很乱的逻辑拆解到各自的小函数里。每个小函数篇幅超过 30 行，继续拆。

Dayuxiaoshui committed:
refactor: split tolerance report generation

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,351 @@
+    return pi
+
+
+def resolve_errno_tolerance(
```
这个函数好。明确 errno 和 tolerance 的区别

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,351 @@
+
+
+def resolve_errno_tolerance(
+    errno2count: dict[int, int], custom_map: dict[int, int] | None
```
custom_map语义不够，改个名字。

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,351 @@
+def calculate_gamma(
+    tolerance: int,
+    pi_value4errno: Callable[[int], float],
+    errno_as_tolerances: dict[int, int],
```
既然类型改成了 dict，那名字就需要换成 errno2tolerance

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,355 @@
+    }
+
+
+def freeze_statistics_at_t1(
```
这个函数语义特别差。t1变量名语义特别模糊。如果本函数只是内部使用，可以加个下划线开头。

lixinqi replied:
应该函数名里的 t1部分弄成参数，比如叫 first_errno_tolerance

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,355 @@
+        }
+
+
+def calculate_parameters_for_tolerance(
```
这个函数语义太弱了。不能从名字看出来它在干什么。

lixinqi replied:
`calculate_es_constructor_params_for_tolerance`
用于计算 es 的那几个变量的值我们统一叫es_constructor_params

lixinqi commented on graph_net/analysis_util.py:
```diff
@@ -514,42 +631,23 @@ def print_stat_info(
-                if not is_correct:
-                    fail_type = "accuracy"
+            # Check correctness using dedicated function
+            is_correct, fail_type = check_sample_correctness(sample, t_key)
```
get_sample_correctness

check_sample_correctness名字给人一种没有返回值，遇到错误就抛异常的柑橘。

lixinqi commented on graph_net/analysis_util.py:
```diff
@@ -558,62 +656,45 @@ def print_stat_info(
+                speedup,
+                fail_type,
+                t_key,
+                is_correct_at_t1[idx],
```
t1 这些变量名语义都不好理解

lixinqi replied:
旧代码一行都不要改。除非是 bug。

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -138,17 +240,31 @@ def main():
-    # 2. Calculate scores for each curve
+    # 2. Calculate scores for each curve and verify aggregated/micro consistency
     all_es_scores = {}
+    tolerance_threshold = 1e-6  # Tolerance for floating point comparison
```
技术报告没有提到需要 1e-6 的默认值。

lixinqi replied:
如果你是想比较宏观和微观两种途径算得的结果，应该穿 es_result_checker 函数，而不是tolerance_threshold。tolerance_threshold不能一下子猜出来它在干什么。而es_result_checker就直截了当定义为：
```python
def es_result_checker(es_from_micro, es_from_macro):
    tol = 1e-3
    return abs(es_from_micro - es_from_macro) < tol
```
我总感觉 1e-6 太大了。

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -138,17 +240,31 @@ def main():
+        # Verify aggregated/micro consistency
+        verified_scores, all_matched = verify_aggregated_micro_consistency(
+            es_scores, folder_name, tolerance_threshold
+        )
```
需要提供一个命令行开关 enable_aggregation_mode，默认开启。

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,108 @@
 from graph_net import analysis_util
 
 
+def compare_single_tolerance_level(
```
名字的语义很弱，看不出来这个函数要做什么

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,351 @@
+from collections.abc import Callable
+
+
+def get_errno_from_error_type(error_type: str) -> int:
```
这个函数名字很好

lixinqi commented on graph_net/samples_statistics.py:
```diff
@@ -0,0 +1,351 @@
+    )
+
+
+def calculate_es_t_from_aggregated(
```
这些名字都非常不错

lixinqi commented on graph_net/verify_aggregated_params.py:
```diff
@@ -0,0 +1,355 @@
+        }
+
+
+def verify_es_components_across_tolerances(
```
`verify_es_constructor_params_across_tolerances`
把本文涉及到的 es_components 都改成es_constructor_params

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,108 @@
+        print(f"\nSUCCESS: All aggregated and micro results match for '{folder_name}'.")
+        print(f"{'='*80}\n")
+
+    return verified_scores, all_matched
```
verified_scores 的语义有点弱。

lixinqi replied:
是不是应该叫 verified_es_values？

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,108 @@
+    all_matched = True
+
+    print(f"\n{'='*80}")
+    print(f"Verifying Aggregated/Micro Consistency for '{folder_name}'")
```
本文所涉的 micro 都应该改成 microscopic。

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,108 @@
+        print(f"{'='*80}\n")
+    else:
+        print(f"\nSUCCESS: All aggregated and micro results match for '{folder_name}'.")
+        print(f"{'='*80}\n")
```
应该直接 assert

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,108 @@
+        print(f"\nSUCCESS: All aggregated and micro results match for '{folder_name}'.")
+        print(f"{'='*80}\n")
+
+    return verified_scores, all_matched
```
返回值应该只需要verified_es_values

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,108 @@
+        )
+
+
+def verify_aggregated_micro_consistency(
```
get_verified_aggregated_es_values

Dayuxiaoshui committed:
refactor: improve naming and semantics for ES calculation

- Rename verify_es_match_at_tolerance to compare_aggregated_es_and_microscopic_es
- Replace tolerance_level with tolerance parameter
- Replace tolerance_threshold with atol/rtol to avoid confusion
- Rename verify_aggregated_microscopic_consistency to get_verified_aggregated_es_values
- Change return type to dict only (remove all_matched)
- Rename verified_scores to verified_es_values
- Replace micro with microscopic throughout
- Rename check_sample_correctness to get_sample_correctness
- Rename t1 variables to first_errno_tolerance
- Rename es_components to es_constructor_params
- Rename calculate_parameters_for_tolerance to calculate_es_constructor_params_for_tolerance
- Rename custom_map to errno_tolerance_overrides
- Rename errno_as_tolerances to errno2tolerance
- Add enable_aggregation_mode command line option

lixinqi commented on graph_net/plot_ESt.py:
```diff
@@ -5,6 +5,134 @@
+    diff = abs(es_from_microscopic - es_from_macro)
+    return diff < atol or diff < rtol * max(
+        abs(es_from_microscopic), abs(es_from_macro), 1e-10
+    )
```
可以调用 np 的 allclose 吗？这些 api 没法一眼看出来谁对谁错

Dayuxiaoshui committed:
feat: add aggregated ES(t) plotting and verification

- Modified plot_ES_results to return fig, ax, all_x_coords for external plotting
- Added manual plotting of aggregated ES(t) curves in main function
- Both microscopic and aggregated curves are plotted on the same graph
- Aggregated curves use dashed lines with square markers for distinction
- All verification checks pass with floating-point precision differences (1.39e-17)

Dayuxiaoshui committed:
fix: move ax.legend outside aggregation condition block

- Move ax.legend() outside the aggregation mode condition block
- Ensure legend is always displayed regardless of aggregation mode
- Fix issue where legend was missing when aggregation mode is disabled

Dayuxiaoshui commented:
<img width="1550" height="952" alt="image" src="https://github.com/user-attachments/assets/f7ecfbfa-8afb-43de-ab5d-ea69177ed5e2" />
